src/Agent.Worker/Build/BuildServer.cs

```diff
@@ -12,22 +11,62 @@
 
 namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
 {
-    public class BuildServer
+    [ServiceLocator(Default = typeof(BuildServer))]
+    public interface IBuildServer : IAgentService
     {
-        private readonly Build2.BuildHttpClient _buildHttpClient;
-        private Guid _projectId;
+        Task ConnectAsync(VssConnection jobConnection);
+        Task<Build2.BuildArtifact> AssociateArtifactAsync(
+            int buildId,
+            Guid projectId,
+            string name,
+            string jobId,
+            string type,
+            string data,
+            Dictionary<string, string> propertiesDictionary,
+            CancellationToken cancellationToken = default(CancellationToken));
+        Task<Build2.Build> UpdateBuildNumber(
+            int buildId,
+            Guid projectId,
+            string buildNumber,
+            CancellationToken cancellationToken = default(CancellationToken));
+        Task<IEnumerable<string>> AddBuildTag(
+            int buildId,
+            Guid projectId,
+            string buildTag,
+            CancellationToken cancellationToken = default(CancellationToken));
+    }
+
+    public class BuildServer : AgentService, IBuildServer
+    {
+        private VssConnection _connection;
+        private Build2.BuildHttpClient _buildHttpClient;
 
-        public BuildServer(VssConnection connection, Guid projectId)
+        public async Task ConnectAsync(VssConnection jobConnection)
         {
-            ArgUtil.NotNull(connection, nameof(connection));
-            ArgUtil.NotEmpty(projectId, nameof(projectId));
+            _connection = jobConnection;
+            int attemptCount = 5;
+            while (!_connection.HasAuthenticated && attemptCount-- > 0)
+            {
+                try
+                {
```

Comment:
We may want to debug log each time we try to connect.

Reply:
We trace when we try and fail, I don't think you'd ever really need more than that.